src/shared/previewUtils.ts

```diff
@@ -0,0 +1,430 @@
+    } else {
+      // all other platforms can work with a base64 PEM text file
+      fileName = path.join(basePath, 'certificate.pem');
+      fs.writeFileSync(fileName, cert.pemCertificate);
+    }
+
+    return fileName;
+  }
+
+  /**
+   * Launches the specified mobile app on the specified device.
+   *
+   * @param platform A mobile platform (iOS or Android)
+   * @param deviceId The identifier (such as name or UDID) of the target device
+   * @param appConfig The app configuration containing info about the mobile app to be launched
+   * @param emulatorPort Optional - only needed when platform is Android and specifies the ADB port of the booted Android virtual device
+   * @param logger An optional logger to be used for logging
+   */
+  public static async launchMobileApp(
+    platform: Platform.ios | Platform.android,
+    appConfig: IOSAppPreviewConfig | AndroidAppPreviewConfig,
+    deviceId: string,
+    emulatorPort?: number,
+    appBundlePath?: string,
+    logger?: Logger
+  ): Promise<void> {
+    logger?.debug(`Attempting to launch mobile app ${appConfig.name}`);
+
+    // If appBundlePath is provided then the app will be installed from
+    // the bundle first then will be launched. Otherwise the assumption
+    // is that app is already installed.
+    if (platform === Platform.ios) {
+      await IOSUtils.launchAppInBootedSimulator(
+        deviceId,
+        appBundlePath,
+        appConfig.id,
+        appConfig.launch_arguments ?? []
+      );
+    } else if (emulatorPort) {
+      // for Android, emulatorPort is required
+      await AndroidUtils.launchAppInBootedEmulator(
+        appBundlePath,
+        appConfig.id,
+        appConfig.launch_arguments ?? [],
+        (appConfig as AndroidAppPreviewConfig).activity,
+        emulatorPort
+      );
+    }
+  }
+
+  /**
+   * Verifies whether a particular app is installed on a mobile device.
+   *
+   * @param platform A mobile platform (iOS or Android)
+   * @param appConfig The app configuration containing info about the mobile app such as name and bundle/package id
+   * @param deviceId The identifier (such as name or UDID) of the target device
+   * @param emulatorPort Optional - only needed when platform is Android and specifies the ADB port of the booted Android virtual device
+   * @param logger An optional logger to be used for logging
+   * @returns `true` if app is installed, `false` otherwise
+   */
+  public static async verifyMobileAppInstalled(
+    platform: Platform.ios | Platform.android,
+    appConfig: IOSAppPreviewConfig | AndroidAppPreviewConfig,
+    deviceId: string,
+    emulatorPort?: number,
+    logger?: Logger
+  ): Promise<boolean> {
+    logger?.debug(`Checking if ${appConfig.id} is installed on device ${deviceId}`);
+    let result = '';
+    try {
+      if (platform === Platform.ios) {
+        result = CommonUtils.executeCommandSync(`xcrun simctl listapps ${deviceId} | grep "${appConfig.id}"`);
+      } else if (emulatorPort) {
```

Comment:
Should we warn or error or something if `emulatorPort` _is_ null? I see from [here](https://github.com/salesforcecli/plugin-lightning-dev/pull/50/files#diff-63422716264f9435db5ca6dcc6618d1c753bff38930fe5490849883d4df7b5a1R99) that the emulator port may not be returned if the device is already booted? Is that going to cause issues in downstream behavior like here?

Reply:
Oh, I'm realizing that this is the platform differentiator again. Can we change this conditional instead to something like:

```typescript
if (platform === Platform.ios) {
  result = CommonUtils.executeCommandSync(`xcrun simctl listapps ${deviceId} | grep "${appConfig.id}"`);
} else {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const resolvedEmulatorPort = emulatorPort!
  result = await AndroidUtils.executeAdbCommand(`shell pm list packages | grep "${appConfig.id}"`, resolvedEmulatorPort);
}
```

This is getting back to my previous comment about separating the behaviors per platform. This is a relatively simple code pattern to follow, and yet the unstated inferences were confusing at first glance.

Reply:
So I'm using `emulatorPort` as a way to differentiator for the platform cases. This will all read much cleaner once I move the functions in this file to `lwc-dev-mobile-core` where we have separate classes/files per platform.

When targeting iOS `emulatorPort` will be `undefined` but when targeting Android it will have a value (even if booting is no-op b/c it is already booted, we still return a valid port).

Reply:
Got it. I wasn't sure the extent of the bigger refactor you had in mind, with these comments. So I figured I'd point them out. Definitely don't spend a lot of time shuffling here, if it's all meant to move below as it were.